### ISSUE TITLE
Unable to create certificates having name more than 64 chars

### ISSUE THREAD

**Issue by ssurenr** [F5Networks/f5-appsvcs-extension#195]
##### Do you already have an issue opened with F5 support?
No

##### ISSUE TYPE

 - Bug Report

##### AS3 BUILD/ VERSION
```
3.13
```

##### BIGIP VERSION

```
BIG-IP 13.1.1.3 Build 0.0.1 Point Release 3
```

##### SUMMARY

```
Any certificate name more than 64 characters long are considered invalid in the declaration. BIG-IP does not have this limitation.
```

##### STEPS TO REPRODUCE
Send the following declaration to BIG-IP

```json
{
    "$schema": "https://raw.githubusercontent.com/F5Networks/f5-appsvcs-extension/master/schema/latest/as3-schema-3.13.0-3.json",
    "class": "AS3",
    "declaration": {
        "suren": {
            "Shared": {
                "class": "Application",
                "surendhar_2018_to_2028_root_certificate_authority_trusted_certificate": {
                    "class": "Certificate",
                    "certificate": "-----BEGIN CERTIFICATE-----\nMIIFmDCCA4ACCQDuuFnZhpzw9zANBgkqhkiG9w0BAQsFADCBjTELMAkGA1UEBhMC\nSU4xEzARBgNVBAgMClRhbWlsIE5hZHUxEjAQBgNVBAcMCVRoYW5qYXZ1cjESMBAG\nA1UECgwJU3VyZW5kaGFyMR8wHQYDVQQDDBZTdXJlbmRoYXIgUmF2aWNoYW5kcmFu\nMSAwHgYJKoZIhvcNAQkBFhFtYWlsQHN1cmVuZGhhci5pbjAeFw0xOTA1MDgwODQy\nNDNaFw0yOTA1MDUwODQyNDNaMIGNMQswCQYDVQQGEwJJTjETMBEGA1UECAwKVGFt\naWwgTmFkdTESMBAGA1UEBwwJVGhhbmphdnVyMRIwEAYDVQQKDAlTdXJlbmRoYXIx\nHzAdBgNVBAMMFlN1cmVuZGhhciBSYXZpY2hhbmRyYW4xIDAeBgkqhkiG9w0BCQEW\nEW1haWxAc3VyZW5kaGFyLmluMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKC\nAgEAxcWMTRN74R/eJxmx0AqFF+dMqj2PgQnKxRnaGm5sEsERK+QWUpSqQ5IJcLeH\nO4KeT3ZsR/COntrzc0trQOH2uevsFA2rSO1Y+aEbIvTTHWeftgK43Dgqds3WfjFr\ngI/kMUaG6vBIVnER2MsNxX5ud0k0YFnj/AKAmWT5VqQbKc9GhJDE6pfTlCCNehko\nEZj7huWvFXOVzUrSrroY8wx750l8l/GvErj38T7cC/BljakTu48BEtZHu/ko4q0R\njgNGsNC2/6QeWB0Ma0BiiDU9mVv0sbR/7s+VvbdVcRwKrYK4gR0xXMJ1BdgDZByw\ncYtRaXoQNonCcuQ5AeLXCoGmqT/bukJE2kcmL6+gIw1ksBF9XKyV9ZYvk+wshO+I\nrjX7QpOfnM9oF5PpkgWtiFssSlvAHsKfHNIu9v4MXuqvYtyWZWg6MWSXmRnAU5ZE\nsJUzB537X+rgx3hQUlMk6GosCPj+pKFfK9dD5e3YaSHuJSvqyNIBE5FtmdnizV7Y\ndS5NrQAFqfUwbUc+wBlwXksG+Xcy1BhG70ev3/Ee86APIIcUPxfhVW0HGGViGWDd\ngYUydsPPU2tZUqpewOUTFnW87k/NzDPC+TxfmAXQrszxBwbJWw8e+iV9U+nAiq6H\nXZTCfANd0EzyxyQPIKEC0p5rTGJHATQ09Y7yW4Se+skb82cCAwEAATANBgkqhkiG\n9w0BAQsFAAOCAgEAgrJHjKVVzorlqMeMzqNbEr+qLA9LbaT3dMrxgYlFd6Ue9TrD\nFrgLIBqeTS5AlZzV5Z5zifOUMjb/BIwpqUzjXOjyFawZsGgMlPI4Ow02zoNsslfk\nSfpd+x+r3sTxOSQ5OalFV5SIg/c+sJqSkZMuwxwD9GPjL54IndZFb0k+vY/NtDtY\n3J5oQz6riu9QywyS3RhDSe0/pTzA5U8nkowx/yM5Fx/lbdPqrXV9CqkkpO+i4Qlo\ncawFZqW0bR0FjoUmnpU8eiUiBWF+M9JeuVMVHf1gTS0RKsXh1VfJtPDrylKWldI5\nXzxAuuzseoBQ9X8eoIoVk2+9oAi/T7UK5HhsLU6d2NVhJAVCr+RRVbYPN4sFhPvk\nu68SIWJErgheUPr8L7BGvsxw2bpJZMLZRGqzNmOyeQnOLckaxWaQPa2bpJAaw7C8\n1eKY3gOgDMTx905z30k2Pn0t4TrWpMs9ttZjBx7rXv6efgOUpnD53dbRAROekdal\nNDIlT8GvChn+QOFnRpXsWGDC2cZks2AIoqI2KjAqLaA6slqdOtvXWv+VJmV13wIa\nUD2bJ+albxmbmFY/lmJOwFIQLeWN817J0y16TXqg0zgzUpfO1ov7S+Ye6sUTq4xj\nrYi0zW5wqHFzI7UBMkFiY0oaHpuimcKPMTG/TgA5bIR3hhGdrzaCuKOsaRU=\n-----END CERTIFICATE-----\n",
									"chainCA": ""
                },
                "template": "shared"
            },
            "class": "Tenant"
        },
        "class": "ADC",
        "id": "urn:uuid:B97DFADF-9F0D-4F6C-8D66-E9B52E593694",
        "label": "CIS Declaration",
        "remark": "Auto-generated by CIS",
        "schemaVersion": "3.11.0"
    }
}
```

##### EXPECTED RESULTS

```
The certificate gets created in the BIG-IP
```

##### ACTUAL RESULTS

```json
{
  "code": 422,
  "errors": [
    "/suren/Shared: should NOT be longer than 64 characters"
  ],
  "declarationFullId": "",
  "message": "declaration is invalid"
}
```

**Comment by ssurenr** [comment]
Also in the error, it is difficult to identify which object does the error talks about.   
Error should indicate the problematic object - `/suren/Shared/surendhar_2018_to_2028_root_certificate_authority_trusted_certificate` instead of just `/suren/Shared/`.

**Comment by dstokesf5** [comment]
The AS3 character limit allows AS3 to generate names based on user input without exceeding the maximum length of a tmsh path. I recommend using a shorter name. If you need more description on the object, consider using the remark or label field.

**Comment by ssurenr** [comment]
@dstokesf5 Certificate name is a variable and it is controlled by the user. Some of our users are service providers and they, in turn, have their customers populating these certificates and the number of certificates that end up in BIG-IP is quite high. 64 is a very short number.

**Comment by dstokesf5** [comment]
AS3 is unlikely to ever match the character limit of BIG-IP. We can ease up the AS3 restrictions a bit by only applying a restriction on the full path instead of each individual element of the path. This would mean a user could go beyond 64 characters as long as their tenant and application names are not also long. I believe this full path restriction would be 195 characters (which includes the forward slashes).

**Comment by ssurenr** [comment]
Sounds like a plan. Also please take care of object name the error response.
https://github.com/F5Networks/f5-appsvcs-extension/issues/195#issuecomment-555134343

**Comment by erjac77** [comment]
Actually, I think it's 255 characters.

**K14440: Folder Path names are limited to 255 characters in length**
https://support.f5.com/csp/article/K14440

Restriction on the full path instead of each individual object would be great, since we often use the FQDN as Tenant/Partition name. 64 chars is pretty short.

**Comment by dstokesf5** [comment]
Like I mentioned earlier, AS3 effectively reserves some of the tmsh character limit to be able to safely generate names by adding appending extra characters onto use supplied names. So the AS3 limit will be less than the tmsh limit of 255.

**Comment by dstokesf5** [comment]
I have created AUTOTOOL-987 for easing the character limit restriction.

**Comment by dstokesf5** [comment]
This was resolved in the 3.17.0 release.
